syntax = "proto3";

service KVMaster {
  rpc GetNodeView(NodeViewRequest) returns (NodeView);
  rpc GetPartitionPrimary(GetPrimaryRequest) returns (GetPrimaryResponse);
}

message NodeViewRequest {}

message NodeView {
  repeated NodeView_Partition partition = 1;
}

message GetPrimaryRequest {
  uint32 partition_idx = 1;
}

message GetPrimaryResponse {
  string addr = 1;  // the primary for your partition is *addr*
  repeated NodeView_Node nodes = 2;  // and here are the other nodes
}

// ==== support ====
message NodeView_Partition {
  int32 primary_idx = 1;
  repeated NodeView_Node node = 2;
}

message NodeView_Node {
  string address = 1;
  bool alive = 2;
}
